### PR TITLE
Made base Denominator constant

### LINKED ISSUE #554: Remove denominators from parameters and declare as constants

### PR DESCRIPTION
**Pull request by adi44**
fixes #554

### CODE REVIEW

File: contracts/Core/StakeManager.sol [hunk 1]
```diff
@@ -312,9 +312,9 @@ contract StakeManager is Initializable, StakeStorage, StateManager, Pause, Stake
         // https://soliditydeveloper.com/stacktoodeep
         {
             (uint16 bountyNum, uint16 burnSlashNum, uint16 keepSlashNum) = (slashNums.bounty, slashNums.burn, slashNums.keep);
-            bounty = (_stake * bountyNum) / baseDenominator;
-            amountToBeBurned = (_stake * burnSlashNum) / baseDenominator;
-            amountToBeKept = (_stake * keepSlashNum) / baseDenominator;
+            bounty = (_stake * bountyNum) / BASE_DENOMINATOR;
+            amountToBeBurned = (_stake * burnSlashNum) / BASE_DENOMINATOR;
+            amountToBeKept = (_stake * keepSlashNum) / BASE_DENOMINATOR;
```
Comment (SkandaBhat) [inline contracts/Core/StakeManager.sol:317]: Does it make sense to optimise this by storing `_stake / BASE_DENOMINATOR`?

**Review by SkandaBhat — CHANGES_REQUESTED** [review]

File: contracts/Core/StakeManager.sol (hunk 1 above)
Comment (adi44) [reply contracts/Core/StakeManager.sol:317]: yeah, you mean to store _stake/BASE_DENOMINATOR  as variable and multiply with corresponding numerator?

variable = _stake/BASE_DENOMINATOR
bounty = variable * bountyNum ...

right? @SkandaBhat

Comment (SkandaBhat) [reply contracts/Core/StakeManager.sol:317]: Yep

File: contracts/Core/storage/Constants.sol [hunk 2]
```diff
@@ -19,6 +19,8 @@ contract Constants {
     uint8 public constant NUM_STATES = 5;
     address public constant BURN_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
 
+    uint16 public constant BASE_DENOMINATOR = 10000;
```
Comment (SkandaBhat) [inline contracts/Core/storage/Constants.sol:22]: Move this to reward manager and state manager

File: contracts/Core/parameters/child/StakeManagerParams.sol [hunk 3]
```diff
@@ -14,12 +14,13 @@ abstract contract StakeManagerParams is ACL, IStakeManagerParams {
...
     uint8 public extendLockPenalty = 1;
     uint8 public maxCommission = 20;
+
     // change the commission by 3% points
     uint8 public deltaCommission = 3;
-    uint16 public override baseDenominator = 10000;
     uint16 public gracePeriod = 8;
     uint16 public epochLength = 300;
     uint16 public epochLimitForUpdateCommission = 100;
+    uint16 public constant BASE_DENOMINATOR = 10000;
```
Comment (hrishikeshio) [inline contracts/Core/parameters/child/StakeManagerParams.sol:23]: constants should be in https://github.com/razor-network/contracts/blob/master/contracts/Core/storage/Constants.sol

**Review by hrishikeshio — CHANGES_REQUESTED** [review]

File: contracts/Core/parameters/child/StakeManagerParams.sol (hunk 3 above)
Comment (adi44) [reply contracts/Core/parameters/child/StakeManagerParams.sol:23]: @hrishikeshio BASE_DENOMINATOR is needed in governance.sol and stakeparams.sol due to which constants were being inherited twice

Comment (adi44) [reply contracts/Core/parameters/child/StakeManagerParams.sol:23]: @hrishikeshio Shall I put it in constants.sol?

Comment (SkandaBhat) [reply contracts/Core/parameters/child/StakeManagerParams.sol:23]: This should be fine. Add a comment here in the code explaining why it needs to be here.

Comment (hrishikeshio) [reply contracts/Core/parameters/child/StakeManagerParams.sol:23]: why is it needed in stakeManagerparams.sol ?

Comment (adi44) [reply contracts/Core/parameters/child/StakeManagerParams.sol:23]: I have added the comment in governance.sol

**Review by abhishekvispute — CHANGES_REQUESTED** [review]
Hi Aditya,
I think we can keep this base denominator in Constants only 
Now on our concern of duplication 

I tried our scenario on remix 

```
pragma solidity 0.8.9;
import './constants.sol';
import  './child.sol';
contract parent is Constants, child{
    
}
```

```
pragma solidity 0.8.9;
import './constants.sol';
contract child is Constants{
    
}
```

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

contract Constants {
    address public constant BURN_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
}
```

now when we compile and check byte code we see 
Its not duplicated, solidity compiler takes care of that 


Bytecode for Constants 
Constants
```
{
	"functionDebugData": {},
	"generatedSources": [],
	"linkReferences": {},
	"object": "608060405234801561001057600080fd5b5060eb8061001f6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063fccc281314602d575b600080fd5b60336047565b604051603e9190609c565b60405180910390f35b73eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee81565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000608882605f565b9050919050565b609681607f565b82525050565b600060208201905060af6000830184608f565b9291505056fea26469706673582212203bcbab48555b9e723d9bc63714e205f09d55a34694548828258130114f7d35b264736f6c63430008090033",
	"opcodes": "PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH2 0x10 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0xEB DUP1 PUSH2 0x1F PUSH1 0x0 CODECOPY PUSH1 0x0 RETURN INVALID PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH1 0xF JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0x4 CALLDATASIZE LT PUSH1 0x28 JUMPI PUSH1 0x0 CALLDATALOAD PUSH1 0xE0 SHR DUP1 PUSH4 0xFCCC2813 EQ PUSH1 0x2D JUMPI JUMPDEST PUSH1 0x0 DUP1 REVERT JUMPDEST PUSH1 0x33 PUSH1 0x47 JUMP JUMPDEST PUSH1 0x40 MLOAD PUSH1 0x3E SWAP2 SWAP1 PUSH1 0x9C JUMP JUMPDEST PUSH1 0x40 MLOAD DUP1 SWAP2 SUB SWAP1 RETURN JUMPDEST PUSH20 0xEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE DUP2 JUMP JUMPDEST PUSH1 0x0 PUSH20 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DUP3 AND SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x88 DUP3 PUSH1 0x5F JUMP JUMPDEST SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x96 DUP2 PUSH1 0x7F JUMP JUMPDEST DUP3 MSTORE POP POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x20 DUP3 ADD SWAP1 POP PUSH1 0xAF PUSH1 0x0 DUP4 ADD DUP5 PUSH1 0x8F JUMP JUMPDEST SWAP3 SWAP2 POP POP JUMP INVALID LOG2 PUSH5 0x6970667358 0x22 SLT KECCAK256 EXTCODESIZE 0xCB 0xAB BASEFEE SSTORE JUMPDEST SWAP15 PUSH19 0x3D9BC63714E205F09D55A34694548828258130 GT 0x4F PUSH30 0x35B264736F6C634300080900330000000000000000000000000000000000 ",
	"sourceMap": "64:109:0:-:0;;;;;;;;;;;;;;;;;;;"
}
```

Bytecode of Parent 
```
{
	"functionDebugData": {},
	"generatedSources": [],
	"linkReferences": {},
	"object": "608060405234801561001057600080fd5b5060eb8061001f6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063fccc281314602d575b600080fd5b60336047565b604051603e9190609c565b60405180910390f35b73eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee81565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000608882605f565b9050919050565b609681607f565b82525050565b600060208201905060af6000830184608f565b9291505056fea264697066735822122096a7045180b8c501d19810e934e53ce3fbf0833fb400c635c2f850c92a5d1fd464736f6c63430008090033",
	"opcodes": "PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH2 0x10 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0xEB DUP1 PUSH2 0x1F PUSH1 0x0 CODECOPY PUSH1 0x0 RETURN INVALID PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH1 0xF JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0x4 CALLDATASIZE LT PUSH1 0x28 JUMPI PUSH1 0x0 CALLDATALOAD PUSH1 0xE0 SHR DUP1 PUSH4 0xFCCC2813 EQ PUSH1 0x2D JUMPI JUMPDEST PUSH1 0x0 DUP1 REVERT JUMPDEST PUSH1 0x33 PUSH1 0x47 JUMP JUMPDEST PUSH1 0x40 MLOAD PUSH1 0x3E SWAP2 SWAP1 PUSH1 0x9C JUMP JUMPDEST PUSH1 0x40 MLOAD DUP1 SWAP2 SUB SWAP1 RETURN JUMPDEST PUSH20 0xEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE DUP2 JUMP JUMPDEST PUSH1 0x0 PUSH20 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DUP3 AND SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x88 DUP3 PUSH1 0x5F JUMP JUMPDEST SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x96 DUP2 PUSH1 0x7F JUMP JUMPDEST DUP3 MSTORE POP POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x20 DUP3 ADD SWAP1 POP PUSH1 0xAF PUSH1 0x0 DUP4 ADD DUP5 PUSH1 0x8F JUMP JUMPDEST SWAP3 SWAP2 POP POP JUMP INVALID LOG2 PUSH5 0x6970667358 0x22 SLT KECCAK256 SWAP7 0xA7 DIV MLOAD DUP1 0xB8 0xC5 ADD 0xD1 SWAP9 LT 0xE9 CALLVALUE 0xE5 EXTCODECOPY 0xE3 0xFB CREATE DUP4 EXTCODEHASH 0xB4 STOP 0xC6 CALLDATALOAD 0xC2 0xF8 POP 0xC9 0x2A 0x5D 0x1F 0xD4 PUSH5 0x736F6C6343 STOP ADDMOD MULMOD STOP CALLER ",
	"sourceMap": "49:35:0:-:0;;;;;;;;;;;;;;;;;;;"
}

```

**Comment by adi44** [comment]
@abhishekvispute  Yeah, even I did research on same, and solidity takes care of it

**Comment by adi44** [comment]
I think we can also put Governance_ROLE in constants too

**Comment by SamAg19** [comment]
> Hi @adi44 , I think we can keep this base denominator in Constants only Now on our concern of duplication
> 
> I tried our scenario on remix
> 
> ```
> pragma solidity 0.8.9;
> import './constants.sol';
> import  './child.sol';
> contract parent is Constants, child{
>     
> }
> ```
> 
> ```
> pragma solidity 0.8.9;
> import './constants.sol';
> contract child is Constants{
>     
> }
> ```
> 
> ```
> // SPDX-License-Identifier: UNLICENSED
> pragma solidity ^0.8.0;
> 
> contract Constants {
>     address public constant BURN_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
> }
> ```
> 
> now when we compile and check byte code we see Its not duplicated, solidity compiler takes care of that
> 
> Bytecode for Constants Constants
> 
> ```
> {
> 	"functionDebugData": {},
> 	"generatedSources": [],
> 	"linkReferences": {},
> 	"object": "608060405234801561001057600080fd5b5060eb8061001f6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063fccc281314602d575b600080fd5b60336047565b604051603e9190609c565b60405180910390f35b73eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee81565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000608882605f565b9050919050565b609681607f565b82525050565b600060208201905060af6000830184608f565b9291505056fea26469706673582212203bcbab48555b9e723d9bc63714e205f09d55a34694548828258130114f7d35b264736f6c63430008090033",
> 	"opcodes": "PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH2 0x10 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0xEB DUP1 PUSH2 0x1F PUSH1 0x0 CODECOPY PUSH1 0x0 RETURN INVALID PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH1 0xF JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0x4 CALLDATASIZE LT PUSH1 0x28 JUMPI PUSH1 0x0 CALLDATALOAD PUSH1 0xE0 SHR DUP1 PUSH4 0xFCCC2813 EQ PUSH1 0x2D JUMPI JUMPDEST PUSH1 0x0 DUP1 REVERT JUMPDEST PUSH1 0x33 PUSH1 0x47 JUMP JUMPDEST PUSH1 0x40 MLOAD PUSH1 0x3E SWAP2 SWAP1 PUSH1 0x9C JUMP JUMPDEST PUSH1 0x40 MLOAD DUP1 SWAP2 SUB SWAP1 RETURN JUMPDEST PUSH20 0xEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE DUP2 JUMP JUMPDEST PUSH1 0x0 PUSH20 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DUP3 AND SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x88 DUP3 PUSH1 0x5F JUMP JUMPDEST SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x96 DUP2 PUSH1 0x7F JUMP JUMPDEST DUP3 MSTORE POP POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x20 DUP3 ADD SWAP1 POP PUSH1 0xAF PUSH1 0x0 DUP4 ADD DUP5 PUSH1 0x8F JUMP JUMPDEST SWAP3 SWAP2 POP POP JUMP INVALID LOG2 PUSH5 0x6970667358 0x22 SLT KECCAK256 EXTCODESIZE 0xCB 0xAB BASEFEE SSTORE JUMPDEST SWAP15 PUSH19 0x3D9BC63714E205F09D55A34694548828258130 GT 0x4F PUSH30 0x35B264736F6C634300080900330000000000000000000000000000000000 ",
> 	"sourceMap": "64:109:0:-:0;;;;;;;;;;;;;;;;;;;"
> }
> ```
> 
> Bytecode of Parent
> 
> ```
> {
> 	"functionDebugData": {},
> 	"generatedSources": [],
> 	"linkReferences": {},
> 	"object": "608060405234801561001057600080fd5b5060eb8061001f6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063fccc281314602d575b600080fd5b60336047565b604051603e9190609c565b60405180910390f35b73eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee81565b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000608882605f565b9050919050565b609681607f565b82525050565b600060208201905060af6000830184608f565b9291505056fea264697066735822122096a7045180b8c501d19810e934e53ce3fbf0833fb400c635c2f850c92a5d1fd464736f6c63430008090033",
> 	"opcodes": "PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH2 0x10 JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0xEB DUP1 PUSH2 0x1F PUSH1 0x0 CODECOPY PUSH1 0x0 RETURN INVALID PUSH1 0x80 PUSH1 0x40 MSTORE CALLVALUE DUP1 ISZERO PUSH1 0xF JUMPI PUSH1 0x0 DUP1 REVERT JUMPDEST POP PUSH1 0x4 CALLDATASIZE LT PUSH1 0x28 JUMPI PUSH1 0x0 CALLDATALOAD PUSH1 0xE0 SHR DUP1 PUSH4 0xFCCC2813 EQ PUSH1 0x2D JUMPI JUMPDEST PUSH1 0x0 DUP1 REVERT JUMPDEST PUSH1 0x33 PUSH1 0x47 JUMP JUMPDEST PUSH1 0x40 MLOAD PUSH1 0x3E SWAP2 SWAP1 PUSH1 0x9C JUMP JUMPDEST PUSH1 0x40 MLOAD DUP1 SWAP2 SUB SWAP1 RETURN JUMPDEST PUSH20 0xEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE DUP2 JUMP JUMPDEST PUSH1 0x0 PUSH20 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DUP3 AND SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x88 DUP3 PUSH1 0x5F JUMP JUMPDEST SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x96 DUP2 PUSH1 0x7F JUMP JUMPDEST DUP3 MSTORE POP POP JUMP JUMPDEST PUSH1 0x0 PUSH1 0x20 DUP3 ADD SWAP1 POP PUSH1 0xAF PUSH1 0x0 DUP4 ADD DUP5 PUSH1 0x8F JUMP JUMPDEST SWAP3 SWAP2 POP POP JUMP INVALID LOG2 PUSH5 0x6970667358 0x22 SLT KECCAK256 SWAP7 0xA7 DIV MLOAD DUP1 0xB8 0xC5 ADD 0xD1 SWAP9 LT 0xE9 CALLVALUE 0xE5 EXTCODECOPY 0xE3 0xFB CREATE DUP4 EXTCODEHASH 0xB4 STOP 0xC6 CALLDATALOAD 0xC2 0xF8 POP 0xC9 0x2A 0x5D 0x1F 0xD4 PUSH5 0x736F6C6343 STOP ADDMOD MULMOD STOP CALLER ",
> 	"sourceMap": "49:35:0:-:0;;;;;;;;;;;;;;;;;;;"
> }
> ```

Doesnt this make issue #528 redundant then? We close that if thats the case

**Review by abhishekvispute — APPROVED** [review]

**Review by SkandaBhat — COMMENTED** [review]
lgtm

**Review by SkandaBhat — APPROVED** [review]

**Pull request merged**